rfcs/0001-reason-message-diff.md

```diff
@@ -0,0 +1,322 @@
+              builder:
+                name: BUILDER-NAME
+                kind: ClusterBuilder
+              cacheSize: "CACHE-SIZE"
+              source:
+                git:
+                  url: GIT-REPOSITORY-URL
+                  revision: GIT-REVISION
+              build:
+                env:
+                  - name: ENV-VAR-NAME
+            -       value: ENV-VAR-VALUE-1
+            +       value: ENV-VAR-VALUE-2                    
+                resources:
+                  limits:
+                    cpu: CPU-LIMIT
+                    memory: MEMORY-LIMIT
+                  requests:
+                    cpu: CPU-REQUEST
+                  memory: MEMORY-REQUEST
+
+Builder:    our-special-url.jfrog.io/instance/cluster-builder:default@sha4398ths9rdpfp9sghbne985-s9ed9hgbd9rz4sh98dxbfs
+Run Image:  index.docker.io/paketobuildpacks/run@sha256:34978twgsui4789et0wh9384erhsg0e798riuhgs0er7iugshe98riugeh99rgaser89
+
+Source:   Local Source 
+
+BUILDPACK ID                            BUILDPACK VERSION
+paketo-buildpacks/bellsoft-liberica     2.13.0
+paketo-buildpacks/maven                 2.3.0
+paketo-buildpacks/executable-jar        2.1.1
+paketo-buildpacks/apache-tomcat         1.5.0
+paketo-buildpacks/dist-zip              1.4.0
+paketo-buildpacks/spring-boot           2.5.0
+```
+
+------------------------------
+
+**For Reason "STACK"**
+
+- Print a diff beneath the reason section
+- Print the stack-name after the reason in order to give context 
+
+------------------------------
+```
+$kp build status petclinic
+Image:      index.docker.io/username/instance@sha256:fsdgh39478th0g9tuisodfhgvns938e54iunwuiehfasf9wgiauwsfx
+Status:     SUCCESS
+Reason:     STACK
+            - RunImage: index.docker.io/paketobuildpacks/run@sha256:34978twgsui4789et0wh9384erhsg0e798riuhgs0er7iugshe98riugeh99rgaser892
+            + RunImage: index.docker.io/paketobuildpacks/run@sha256:fh34578tg29wb645rwhs87e5hrw0e98jtfnws9e7w378qe49tr3g39n5gwe9u5hge549w
+            
+Builder:    our-special-url.jfrog.io/instance/cluster-builder:default@sha4398ths9rdpfp9sghbne985-s9ed9hgbd9rz4sh98dxbfs
+Run Image:  index.docker.io/paketobuildpacks/run@sha256:fh34578tg29wb645rwhs87e5hrw0e98jtfnws9e7w378qe49tr3g39n5gwe9u5hge549w
+
+Source:   Local Source 
+
+BUILDPACK ID                            BUILDPACK VERSION
+paketo-buildpacks/bellsoft-liberica     2.13.0
+paketo-buildpacks/maven                 2.3.0
+paketo-buildpacks/executable-jar        2.1.1
+paketo-buildpacks/apache-tomcat         1.5.0
+paketo-buildpacks/dist-zip              1.4.0
+paketo-buildpacks/spring-boot           2.5.0
+```
+
+------------------------------
+
+**For Reason "BUILDPACK"**
+
+- Print a diff as part of the buildpacks table 
+
+------------------------------
+
+```
+$kp build status petclinic
+Image:      index.docker.io/username/instance@sha256:fsdgh39478th0g9tuisodfhgvns938e54iunwuiehfasf9wgiauwsfx
+Status:     SUCCESS
+Reason:     BUILDPACK
```

Comment:
Will this buildpack diff account for both language family as well as implementation buildpack updates? Currently, kp doesn't show the tree-like buildpack output (similar to `pack inspect-builder`), which makes it really hard to figure out specific buildpacks/dependencies that were updated.

Also, any thoughts on the diff showing the actual dependencies that changed? This would let users determine whether a specific buildpack update is due to upstream CVE, new runtime, etc. (dependency update) or a feature/bug fix in the buildpack itself.

Reply:
> Will this buildpack diff account for both language family as well as implementation buildpack updates? Currently, kp doesn't show the tree-like buildpack output (similar to pack inspect-builder), which makes it really hard to figure out specific buildpacks/dependencies that were updated.

The rebuild logic is agnostic to the meta buildpack hierarchy. It simplistically will schedule a rebuild when the builder no longer has an implementation buildpack that was previously used. So, this output should likely represent that simplicity.  However, `kp builder status` should be updated to show the tree like output.

Reply:
> Also, any thoughts on the diff showing the actual dependencies that changed? This would let users determine whether a specific buildpack update is due to upstream CVE, new runtime, etc. (dependency update) or a feature/bug fix in the buildpack itself.

At this point this diff view largely shows *why* a build was scheduled as a window into the kpack declarative logic. It seems like some other diff-like functionality would be better positioned to show relevant dependency changes. 

However, surfacing this info about what dependency changes happen and why they might matter (CVE, bug fix, etc) somewhere would be very interesting and would require some more buildpack metadata.

Reply:
Thanks, that makes sense! Happy to chat about what changes we'd need to make in the buildpacks to support this.